Test stake-pool metrics on extremely short epochs

Anviking opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#713 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] ⚠️ I have changed the epoch length to 3, and slotDuration to 1, in the genesis block of the integration tests.
- [x] I have extended the expectation that we `eventually get 200`, to `eventually get 200 >> eventually get 503`, as the stake-pool metrics become unsynced when the node enters a new epoch.
- [x] I have added guarantees on the returned list of stake-pools based on our specific testing setup. 

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

piotr-iohk commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -38,6 +47,25 @@ spec = do
+        -- to test with very short epochs.
+        eventually $ do
+            r <- request @[ApiStakePool] ctx listStakePoolsEp Default Empty
+            expectResponseCode HTTP.status503 r
```
Would be nice to have a nice message returned here explaining why it is `503`...

KtorZ replied:
We need to make sure that the polling delay in eventually is short in front of the stepping delay of the networking layer, otherwise, it might take a lot of tries :/ 

Also, on a different note, this should really be another test scenario.

Anviking replied:
There is a message, if that's what you mean! (see gifs in description of https://github.com/input-output-hk/cardano-wallet/pull/881)

```
I can't list stake pools yet because I need to scan the blockchain for metrics first. I'm at 0%
```

*Could maybe check that the message ends with a "%", to keep it simple* 🤔

piotr-iohk replied:
Good! Then you can use:
```
expectErrorMessage "I can't list stake pools yet because I need to scan the blockchain for metrics first. I'm at"
```
(it actually checks if message contains text, not if the message is exactly like ...)

Anviking replied:
> Also, on a different note, this should really be another test scenario.

I wasn't sure the same jormungandr instance was used for all the scenarios, but I see it is. Then it makes sense 👍 (I want to make sure it returns 200 _before_ it returns 503 again)

KtorZ commented on lib/jormungandr/test/data/jormungandr/genesis.yaml:
```diff
@@ -20,14 +20,14 @@ blockchain_configuration:
 
   # Number of slots in each epoch
-  slots_per_epoch: 500
+  slots_per_epoch: 3
```
3 seems like.. a bit excessive :/ but I have no strong argument to favor `10` or `500` more than `3` ...

Anviking replied:
I think being excessive here is the only way we can make the test reliably succeed in a short enough time.

I agree, it feels a bit dangerous…

KtorZ commented on lib/core-integration/src/Test/Integration/Framework/DSL.hs:
```diff
@@ -762,6 +766,27 @@ syncProgress =
+    :: HasType (Quantity "block" Natural) s
+    => Lens' s (Quantity "block" Natural)
+blocks =
+    lens (view typed) (\(s, v) -> set typed v s)
```
These lenses are somewhat pointless here :grimacing: ... The other lenses in this file are usually dealing with type conversions to actually makes it easier to write assertions after. For instance, this one would be better defined as:

```hs
blocks
    :: HasType (Quantity "block" Natural) s
    => Lens' s Natural
```

Anviking replied:
That specifically doesn't work because it complains about overlapping instances. I believe because of the following, already existing, definition:
```haskell
amount :: HasType (Quantity "lovelace" Natural) s => Lens' s Natural
```

So I wrote it using `HasField'` instead:
```haskell
stake
    :: HasField' "controlledStake" s (Quantity "lovelace" Natural)
    => Lens' s Natural
stake = lens
    (getQuantity . getField @"controlledStake")
    (\(s, v) -> setField @"controlledStake" (Quantity v) s)

blocks
    :: HasField' "producedBlocks" s (Quantity "block" Natural)
    => Lens' s Natural
blocks = lens
    (getQuantity . getField @"producedBlocks")
    (\(s, v) -> setField @"producedBlocks" (Quantity v) s)
```

Anviking committed:
Change integration test epoch length to 3

Anviking committed:
Stronger expectations on /stake-pool integration tests

- expectations on 503 error message
- expectations on 200 data specific to our setup
- expectation that we will 503 eventually on epoch boundaries

Anviking commented on lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:
```diff
@@ -38,8 +46,35 @@ spec = do
+                    \blockchain for metrics first. I'm at"
+                ]
+
+    describe "STAKE_POOLS_LIST_03 - v2/stake-pools - Methods Not Allowed" $ do
```
@piotr-iohk This ended up being re-numbered from 02 to 03 because I liked the order better. Ok, or should the numbers be preserved?

piotr-iohk replied:
No worries. It is just to roughly match the capabilities in the [doc](https://docs.google.com/spreadsheets/d/1fadIA_j4nCd3FbylPo5J8K9Fy6tixSC2T9V3xHKuUvU/edit#gid=0), to make sure everything is covered. But actually didn't put anything for STAKE_POOLS yet, but even if I did, that's not an issue cause I'd reshuffle it later on.

piotr-iohk reviewed (approved):
:+1:
